plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/pages/htmlPreprocessors.kt

```diff
@@ -89,74 +89,89 @@ object IndexGenerator : PageTransformer {
         }
         val keys = elements.keys.sortedBy { it }
         val sortedElements = elements.entries.sortedBy { (a, _) -> a }
-        return input.modified(children = input.children + sortedElements.mapIndexed { i, (_, set) ->
-            IndexPage(i + 1, set.sortedBy { it.getId().toLowerCase() }, keys, input.sourceSets())
-        })
+
+        val indexNodeComparator = compareBy<NavigableJavadocNode> { it.getId().toLowerCase() }
+            .thenBy { it.getFullComparatorKey() }
+
+        val indexPages = sortedElements.mapIndexed { idx, (_, set) ->
+            IndexPage(
+                id = idx + 1,
+                elements = set.sortedWith(indexNodeComparator),
+                keys = keys,
+                sourceSet = input.sourceSets()
+            )
+        }
+        return input.modified(children = input.children + indexPages)
+    }
+
+    private fun NavigableJavadocNode.getFullComparatorKey(): String {
+        return getDRI().let { dri ->
+            val packageName = dri.packageName.orEmpty()
+            val className = dri.classNames.orEmpty()
+            val callableName = dri.callable?.name.orEmpty()
+            val parameters = dri.callable?.signature().orEmpty()
+
+            "$packageName/$className/$callableName/$parameters"
+        }
     }
 }
 
 object DeprecatedPageCreator : PageTransformer {
     override fun invoke(input: RootPageNode): RootPageNode {
-        val elements = HashMap<DeprecatedPageSection, MutableSet<DeprecatedNode>>().apply {
-
-            fun <T> T.putAs(deprecatedPageSection: DeprecatedPageSection) where
-                    T : NavigableJavadocNode,
-                    T : WithJavadocExtra<out Documentable> {
-                val deprecatedNode = DeprecatedNode(
-                    listOfNotNull(
-                        getDRI().packageName?.takeUnless { it.isBlank() },
-                        if (this is JavadocFunctionNode) getAnchor() else getId()
-                    ).joinToString("."),
-                    getDRI(),
-                    (this as? WithBrief)?.brief.orEmpty()
-                )
-                getOrPut(deprecatedPageSection) { mutableSetOf() }.add(deprecatedNode)
-                if (deprecatedAnnotation?.params?.get("forRemoval") == BooleanValue(true)) {
-                    getOrPut(DeprecatedPageSection.DeprecatedForRemoval) { mutableSetOf() }.add(deprecatedNode)
-                }
+        val elements = linkedMapOf<DeprecatedPageSection, LinkedHashSet<DeprecatedNode>>()
```

Comment:
The only thing changed here is `HashMap` -> `linkedMapOf()` and `MutableSet` -> `LinkedHashSet`. The rest comes from removing `.apply {}` from the HashMap